<a href="https://colab.research.google.com/github/vhpraneeth/VIT_DWDM/blob/main/Lab_9_DWDM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

18BCE7147

[GitHub Link](https://github.com/vhpraneeth/VIT_DWDM/blob/main/Lab_9_DWDM.ipynb)

## Ensemble Methods 


# Boosting 



# AdaBoost / Adaptive Boosting





In [1]:
from sklearn.ensemble import AdaBoostClassifier
ada_clf = AdaBoostClassifier()
ada_clf.fit(X_train, y_train)

NameError: ignored

In [ ]:
print_score(ada_clf, X_train, X_test, y_train, y_test, train=True)


Train Result:

accuracy score: 0.8896

Classification Report: 
               precision    recall  f1-score   support

           0       0.87      0.77      0.82        53
           1       0.90      0.95      0.92       110

    accuracy                           0.89       163
   macro avg       0.88      0.86      0.87       163
weighted avg       0.89      0.89      0.89       163


Confusion Matrix: 
 [[ 41  12]
 [  6 104]]

ROC AUC: 0.8595

Average Accuracy: 	 0.7441
Accuracy SD: 		 0.0834


In [ ]:
print_score(ada_clf, X_train, X_test, y_train, y_test, train=False)

Test Result:

accuracy score: 0.7895

Classification Report: 
               precision    recall  f1-score   support

           0       0.67      0.67      0.67         6
           1       0.85      0.85      0.85        13

    accuracy                           0.79        19
   macro avg       0.76      0.76      0.76        19
weighted avg       0.79      0.79      0.79        19


Confusion Matrix: 
 [[ 4  2]
 [ 2 11]]

ROC AUC: 0.7564



## AdaBoost with Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
ada_clf = AdaBoostClassifier(RandomForestClassifier())
ada_clf.fit(X_train, y_train)

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=RandomForestClassifier(bootstrap=True,
                                                         ccp_alpha=0.0,
                                                         class_weight=None,
                                                         criterion='gini',
                                                         max_depth=None,
                                                         max_features='auto',
                                                         max_leaf_nodes=None,
                                                         max_samples=None,
                                                         min_impurity_decrease=0.0,
                                                         min_impurity_split=None,
                                                         min_samples_leaf=1,
                                                         min_samples_split=2,
                                       

In [ ]:
print_score(ada_clf, X_train, X_test, y_train, y_test, train=True)

Train Result:

accuracy score: 0.9387

Classification Report: 
               precision    recall  f1-score   support

           0       0.91      0.91      0.91        53
           1       0.95      0.95      0.95       110

    accuracy                           0.94       163
   macro avg       0.93      0.93      0.93       163
weighted avg       0.94      0.94      0.94       163


Confusion Matrix: 
 [[ 48   5]
 [  5 105]]

ROC AUC: 0.9301

Average Accuracy: 	 0.7985
Accuracy SD: 		 0.0852


In [ ]:
print_score(ada_clf, X_train, X_test, y_train, y_test, train=False)

Test Result:

accuracy score: 0.8421

Classification Report: 
               precision    recall  f1-score   support

           0       0.80      0.67      0.73         6
           1       0.86      0.92      0.89        13

    accuracy                           0.84        19
   macro avg       0.83      0.79      0.81        19
weighted avg       0.84      0.84      0.84        19


Confusion Matrix: 
 [[ 4  2]
 [ 1 12]]

ROC AUC: 0.7949



# Gradient Boosting / Gradient Boosting Machine (GBM)

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
gbc_clf = GradientBoostingClassifier()
gbc_clf.fit(X_train, y_train)

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='deprecated',
                           random_state=None, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [ ]:
print_score(gbc_clf, X_train, X_test, y_train, y_test, train=True)

Train Result:

accuracy score: 0.9325

Classification Report: 
               precision    recall  f1-score   support

           0       0.94      0.85      0.89        53
           1       0.93      0.97      0.95       110

    accuracy                           0.93       163
   macro avg       0.93      0.91      0.92       163
weighted avg       0.93      0.93      0.93       163


Confusion Matrix: 
 [[ 45   8]
 [  3 107]]

ROC AUC: 0.9109

Average Accuracy: 	 0.7982
Accuracy SD: 		 0.0815


In [ ]:
print_score(gbc_clf, X_train, X_test, y_train, y_test, train=False)

Test Result:

accuracy score: 0.8421

Classification Report: 
               precision    recall  f1-score   support

           0       0.80      0.67      0.73         6
           1       0.86      0.92      0.89        13

    accuracy                           0.84        19
   macro avg       0.83      0.79      0.81        19
weighted avg       0.84      0.84      0.84        19


Confusion Matrix: 
 [[ 4  2]
 [ 1 12]]

ROC AUC: 0.7949



| Classifier | Decision Tree | Bagging | Random Forest | Optimised RF | Extra-Trees | AdaBoost (CART) | AdaBoost (RF) | Gradient Boosting |
|:-|:-|:- |:- |:- |:- |:-|:-| :- |
| Train accuracy score | 0.9528 | 0.9528 | 0.9325 | 0.9264 | 0.9448 | 0.8661 | 0.9528 | 0.9449 |
| Average accuracy score | 0.7724 | 0.7879 | 0.7801 | 0.7059 | 0.7548 | 0.7793 | 0.7353 | 0.7906 |
| SD | 0.1018 | 0.1008 | 0.1474 | 0.1308 | 0.1406 | 0.1172 | 0.0881 | 0.0912 |
| Test accuracy score | 0.7636 | 0.7455 | 0.7895 | 0.6316 | 0.7895 | 0.6545 | 0.7818 | 0.7818 |


# XGBoost (Extreme Gradient Boosting)



In [ ]:
import xgboost as xgb

In [ ]:
xgb_clf = xgb.XGBClassifier(max_depth=5, n_estimators=10000, learning_rate=0.3,
                            n_jobs=-1)

In [ ]:
xgb_clf.fit(X_train, y_train)

In [ ]:
print_score(xgb_clf, X_train, y_train, X_test, y_test, train=True)

In [ ]:
print_score(xgb_clf, X_train, y_train, X_test, y_test, train=False)